src/main/java/uk/gov/pay/directdebit/mandate/services/gocardless/GoCardlessMandateStateCalculator.java

```diff
@@ -51,14 +93,37 @@
                             GOCARDLESS_ACTIONS_THAT_CHANGE_STATE);
                 })
                 .orElse(Optional.empty());
+    }
+
+    private DirectDebitStateWithDetails<MandateState> mapEventToState(Event event, Mandate mandate) {
+        if (event instanceof GoCardlessEvent) {
+            return mapGoCardlessEventToState((GoCardlessEvent) event, mandate);
+        } else if (event instanceof GovUkPayEvent) {
+            return mapGovUkPayEventToState((GovUkPayEvent) event, mandate);
+        } else {
+            throw new IllegalArgumentException(format("Unexpected Event of type %s", event.getClass()));
+        }
+    }
```

Comment:
We should consider reintroducing an enum for GoCardless event actions again. Then we could have a common interface for that and `GovUkPayEventType` and use a single map for mapping them to events, avoiding the need for this crime against polymorphism 😛

Reply:
Added a method on the `Event` interface so we don't have to do this.